### PR TITLE
Use esp-mqtt from esp-idf

### PR DESCRIPTION
**Pull request by ryanplusplus**
Addresses https://github.com/nodemcu/nodemcu-firmware/pull/2501#issuecomment-427643017

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/en/*`.

Per @devsaurus in https://github.com/nodemcu/nodemcu-firmware/pull/2501#issuecomment-427643017, esp-idf now contains esp-mqtt so we no longer need to include this as a submodule. This work removes the old submodule and updates the implementation of the mqtt module to be compatible with the newer version of esp-mqtt.

### CODE REVIEW

File: components/modules/mqtt.c [hunk 1]
```diff
@@ -8,6 +8,24 @@
...
+typedef struct {
+	esp_mqtt_client_config_t config;
+	char host[MQTT_MAX_HOST_LEN];
+	char uri[MQTT_MAX_HOST_LEN];
+	char client_id[MQTT_MAX_CLIENT_LEN];
+	char username[MQTT_MAX_USERNAME_LEN];
+	char password[MQTT_MAX_PASSWORD_LEN];
+	char lwt_topic[MQTT_MAX_LWT_TOPIC];
+	char lwt_msg[MQTT_MAX_LWT_MSG];
+} esp_mqtt_lua_client_config_t;
```
Comment (ryanplusplus) [inline components/modules/mqtt.c]: The older version of esp-mqtt included character arrays in `esp_mqtt_client_config_t`. The new version includes a `const char *` to a user-allocated string. In order to add compatibility with the new configuration type I defined a new parent struct for the config that allocates the memory that used to be allocated by `esp_mqtt_client_config_t` and pointed the new fields to that memory.

Comment (devsaurus) [reply components/modules/mqtt.c]: Elegant solution :wink:

File: components/modules/mqtt.c [hunk 2]
```diff
@@ -634,13 +652,20 @@ static int mqtt_new( lua_State* L )
 	clientid = luaL_checkstring( L, 1 );
 	NODE_DBG("MQTT client id %s\n", clientid);
 
-	esp_mqtt_client_config_t * mqtt_cfg = (esp_mqtt_client_config_t *) malloc(sizeof(esp_mqtt_client_config_t));
-	memset(mqtt_cfg, 0, sizeof(esp_mqtt_client_config_t));
+	esp_mqtt_lua_client_config_t * mqtt_cfg = (esp_mqtt_lua_client_config_t *) malloc(sizeof(esp_mqtt_lua_client_config_t));
+	memset(mqtt_cfg, 0, sizeof(esp_mqtt_lua_client_config_t));
+	mqtt_cfg->config.host = mqtt_cfg->host;
```
Comment (devsaurus) [inline components/modules/mqtt.c]: When compiling locally, gcc complains:
> components/modules/mqtt.c:657:24: error: assignment to expression with array type
>  mqtt_cfg->config.host = mqtt_cfg->host;

No clue why the [travis build](https://travis-ci.org/nodemcu/nodemcu-firmware/builds/438341577?utm_source=github_status&utm_medium=notification) passes without issues.

**Comment by devsaurus** [comment]
This PR seems to re-wind the submodule in `components/u8g2/u8g2`. Can you please rebase onto current dev-esp32 branch?

**Comment by ryanplusplus** [comment]
@devsaurus sorry about that, should be good now.

**Comment by ryanplusplus** [comment]
Thanks!

**Comment by devsaurus** [comment]
Thank **you** for supporting this project!

**Pull request merged**